src/spec/helpers/iterate.cr

```diff
@@ -0,0 +1,99 @@
+module Spec::Methods
+  # Spec helper for generic iteration methods which tests both yielding and
+  # iterator overloads.
+  #
+  # This helper creates two spec examples named *description* with suffixes
+  # `" yielding"` and `" iterator"`.
+  # The yielding example calls *method* with a block and expects the iteration
+  # elements to be yielded to the block. The iterator example calls *method*
+  # without a block and expects it to return an `Iterator` which it then consumes.
+  #
+  # The iterated elements are collected in an array and compared to *expected*,
+  # ensuring type-equality of the elements.
+  #
+  # By default, both examples make sure that the iteration is finished after
+  # iterating all elements from *expected*. If the iteration is infinite,
+  # passing `infinite: true` skips that check and allows to test a finite sample
+  # of an infinite iteration.
+  #
+  # ```
+  # require "spec/helpers/iterate"
+  #
+  # it_iterates "Array#each", [1, 2, 3], (1..3).each
+  # it_iterates "infinite #cycle", [1, 2, 3, 1, 2, 3, 1], (1..3).cycle, infinite: true
+  # ```
+  #
+  # If the iteration elements are tuples (i.e. multiple values), the yielding
+  # variant by default only catches the first value because of the block argument
+  # mechanics. Passing `tuple: true` ensures all yielded arguments are collected
+  # using a splat.
+  #
+  # ```
+  # require "spec/helpers/iterate"
+  #
+  # it_iterates "Array#each_with_index", [{1, 0}, {2, 1}, {3, 2}], (1..3).each_with_index, tuple: true
+  # ```
+  macro it_iterates(description, expected, method, *, infinite = false, tuple = false)
+    it "{{ description.id }} yielding" do
+      assert_iterates_yielding {{ expected }}, {{ method }}, infinite: {{ infinite }}, tuple: {{ tuple }}
+    end
+    it "{{ description.id }} iterator" do
+      assert_iterates_iterator {{ expected }}, {{ method }}, infinite: {{ infinite }}
+    end
+  end
+
+  # Calls *method* with a block and compares yielded values with *expected*.
+  #
+  # See `.it_iterates` for details.
+  macro assert_iterates_yielding(expected, method, *, infinite = false, tuple = false)
+    %remaining = ({{expected}}).size
+    %ary = [] of typeof(({{ expected }})[0])
```

Comment:
Not needed now, but maybe we can use the new element_type in case someone passes a tuple here.